torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -866,6 +866,72 @@ std::vector<xla::XlaOp> BuildSgdOptimizerStep(
   return results;
 }
 
+std::vector<xla::XlaOp> BuildAdamOptimizerStep(
+    const xla::XlaOp& found_inf, const xla::XlaOp& step,
+    const xla::XlaOp& param, const xla::XlaOp& grad, const xla::XlaOp& exp_avg,
+    const xla::XlaOp& exp_avg_sq, const xla::XlaOp& max_exp_avg_sq,
+    const xla::XlaOp& beta1, const xla::XlaOp& beta2, const xla::XlaOp& lr,
+    const xla::XlaOp& weight_decay, const xla::XlaOp& eps,
+    bool use_weight_decay, bool use_amsgrad, bool use_adamw) {
+  // XLA version of Adam/AdamW algorithm
+  // https://github.com/pytorch/pytorch/blob/master/torch/optim/_functional.py#L64-L110
+  // https://github.com/pytorch/pytorch/blob/master/torch/optim/_functional.py#L112-L155
+
+  xla::PrimitiveType type = XlaHelpers::ShapeOfXlaOp(param).element_type();
+  xla::XlaOp one = xla::One(param.builder(), type);
+  xla::XlaOp zero = xla::Zero(param.builder(), type);
+
+  xla::XlaOp found_inf_cond = xla::Ne(found_inf, zero);
+
+  // update step counter first
+  xla::XlaOp not_found_inf =
+      xla::ConvertElementType(xla::Not(found_inf_cond), type);
+  xla::XlaOp new_step = step + not_found_inf;
```

Comment:
Could you add a comment here to explain this `new_step` computation?

Reply:
Done.